### PR TITLE
Spin lock deadlock detection

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

File: core/arch/arm/include/kernel/spinlock.h [hunk 1]
```diff
@@ -57,13 +57,46 @@ void __cpu_spin_unlock(unsigned int *lock);
...
+#ifdef CFG_TEE_CORE_DEBUG
+#define cpu_spin_lock(lock) \
+	cpu_spin_lock_dldetect(__func__, __LINE__, lock)
+
+static inline void cpu_spin_lock_dldetect(const char *func, const int line,
+					  unsigned int *lock)
+{
+	unsigned int retries = 0;
+	unsigned int reminder = 0;
+	assert(thread_foreign_intr_disabled());
```
Comment (etienne-lms) [inline core/arch/arm/include/kernel/spinlock.h]: minor: empty line before the assert.

File: core/arch/arm/mm/tee_pager.c [hunk 2]
```diff
@@ -189,10 +189,38 @@ static tee_mm_entry_t *pager_alias_area;
...
+#define pager_lock(ai) pager_lock_dldetect(__func__, __LINE__, ai)
+
+static uint32_t pager_lock_dldetect(const char *func, const int line,
+				    struct abort_info *ai)
+{
+	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	unsigned int retries = 0;
+	unsigned int reminder = 0;
+
+	while (!cpu_spin_trylock(&pager_spinlock)) {
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c]: `!` or not `!` ?

Comment (igoropaniuk) [inline core/arch/arm/mm/tee_pager.c]:
```
/* returns 0 on locking success, non zero on failure */
unsigned int __cpu_spin_trylock(unsigned int *lock);
....
static inline bool cpu_spin_trylock(unsigned int *lock)
{
	unsigned int rc;

	assert(thread_foreign_intr_disabled());
	rc = __cpu_spin_trylock(lock);
	if (!rc)
		spinlock_count_incr();
	return !rc;
}
```

Comment (igoropaniuk) [inline core/arch/arm/mm/tee_pager.c]: I don't actually understand why  `cpu_spin_trylock() == !__cpu_spin_trylock()`, which can throw into confusion

Comment (jenswi-linaro) [inline core/arch/arm/mm/tee_pager.c]: I agree, it is confusing. It stems from `strex` returning `1` in `<Rd>` on failure and `0` on success and that `__cpu_spin_trylock()` tries to do the bare minimum.

The code here is correct though.

**Comment by jenswi-linaro** [comment]
Review comment addressed

**Comment by jenswi-linaro** [comment]
Tag applied

**Pull request merged**